pep-0572.rst

```diff
@@ -275,16 +275,47 @@ i`` part establishes that ``i`` is local to the comprehension, but the
 ``i :=`` part insists that ``i`` is not local to the comprehension.
 The same reason makes these examples invalid too::
 
-    [[(j := j) for i in range(5)] for j in range(5)]
-    [i := 0 for i, j in stuff]
-    [i+1 for i in i := stuff]
+    [[(j := j) for i in range(5)] for j in range(5)] # compiler TargetScopeError
+    [i := 0 for i, j in stuff]                       # compiler TargetScopeError
+    [i+1 for i in (i := stuff)]                      # compiler TargetScopeError
+
+While it's technically possible to assign consistent semantics to these cases,
+it's difficult to determine whether those semantics actually make *sense* in the
+absence of real use cases. Accordingly, the reference implementation will ensure
+that such cases raise ``TargetScopeError``, a new subclass of ``SyntaxError``
+(with the same signature), rather than executing with implementation defined
+behaviour.
+
+This restriction applies even if the assignment expression is never executed::
+
+    [False and (i := 0) for i, j in stuff]     # compiler TargetScopeError
+    [i for i, j in stuff if True or (j := 1)]  # compiler TargetScopeError
+
+For the comprehension body (the part before the first "for" keyword) and the
+filter expression (the part after "if" and before any nested "for"), this
+restriction applies solely to target names that are also used as iteration
+variables in the comprehension. Lambda expressions appearing in these
+positions introduce a new explicit function scope, and hence may use assignment
+expressions with no additional restrictions.
+
+Due to design constraints in the reference implementation, named expressions
+are disallowed entirely as part of comprehension iterable expressions (the part
+after "in", and before any subsequent "if" or "for" keyword)::
```

Comment:
after the first "in"

Reply:
I went back and forth on whether or not to allow assignment expressions in nested iterable expressions, and ended up settling on "Don't allow them at all", as that seemed like a simpler restriction to explain than "They're only restricted by name, except for the leftmost one, where they're not allowed at all" (plus the general principle of "It's easier to relax a restriction later than it is to tighten it").

Flipping between the two variants in the draft implementation PR is pretty straightforward: to allow them in inner loops, we would change `symtable_visit_comprehension` to skip incrementing the new `ste_comp_iter_expr` counter.

Reply:
I've clarified the wording and the examples to better match the current PR, so I'm going to mark this particular thread as resolved, but as noted above, it would be quite feasible to make the restriction less severe than it is in the current PR.